File: Entities/OmniZipMover.cs
using Celeste.Mod.Entities;
using System.Collections;
using System.Collections.Generic;
using System;
using ChroniaHelper.Utils;
using ChroniaHelper.Cores;
using System.Runtime.CompilerServices;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace ChroniaHelper.Entities;

[Tracked(false)]
[CustomEntity("ChroniaHelper/OmniZipMover")]
public class OmniZipMover : OmniZipSolid
{

    public enum Themes
    {
        Normal,
        Moon,
        Cliffside
    }
    public readonly Themes theme;

    private readonly MTexture[,] edges = new MTexture[3, 3];
    private readonly MTexture[,] innerCorners = new MTexture[2, 2];
    
    private readonly List<MTexture> innerCogs;
    private readonly MTexture temp = new();

    private readonly BloomPoint bloom;

    private readonly Coroutine seq;


    public OmniZipMover(EntityData data, Vector2 offset)
        : this(data, data.Position + offset, data.Width, data.Height, data.NodesWithPosition(offset),
              data.Enum("theme", Themes.Normal),
              data.Bool("permanent"),
              data.Bool("waiting"),
              data.Bool("ticking"),
              data.Attr("customSkin").Trim(),
              data.Attr("backgroundColor").Trim(),
              data.Attr("ropeColor").Trim(),
              data.Attr("ropeLightColor").Trim(),
              data.Attr("customBlockTexture").Trim(),
              data.Attr("delays").Trim(),
              data.Attr("easing","sinein"),
              data.Attr("nodeSpeeds").Trim(),
              data.Int("ticks"),
              data.Float("tickDelay"),
              data.Bool("synced"),
              data.Float("startDelay"),
              data.Attr("startSound", "event:/CommunalHelperEvents/game/zipMover/normal/start"),
              data.Attr("impactSound", "event:/CommunalHelperEvents/game/zipMover/normal/impact"),
              data.Attr("tickSound", "event:/CommunalHelperEvents/game/zipMover/normal/tick"),
              data.Attr("returnSound", "event:/CommunalHelperEvents/game/zipMover/normal/return"),
              data.Attr("finishSound", "event:/CommunalHelperEvents/game/zipMover/normal/finish"),
              data.Bool("customSound",false),
              data.Bool("startShaking", true),
              data.Bool("nodeShaking",true),
              data.Bool("returnShaking",true),
              data.Bool("tickingShaking",true),
              data.Bool("permanentArrivalShaking",true),
              data.Attr("syncTag",""),
              data.Bool("hideRope",false),
              data.Bool("hideCog",false),
              data.Bool("dashable", false),
              data.Bool("onlyDashActivate", false),
              data.Bool("dashableOnce", true),
              data.Bool("dashableRefill", false),
              data.Attr("returnDelays", "0.2,0.2"),
              data.Attr("returnSpeeds","1,1"),
              data.Float("returnedIrrespondingTime", 0.5f),
              data.Attr("returnEasing", "sinein"),
              data.Bool("nodeSound",true),
              data.Bool("timeUnits", false),
              data.Bool("sideFlag",false)
              )
    { }

    public OmniZipMover(EntityData data, Vector2 position,
        int width, int height, Vector2[] nodes,
        Themes theme, bool permanent, bool waits,
        bool ticking, string customSkin,
        //新增变量
        string bgColor, string ropeColor, string ropeLightColor,
        string legacyCustomTexture, string delays, string ease,
        string speeds, int ticks, float tickDelay, bool sync, float startdelay,
        string startSfx, string impactSfx, string tickSfx, string returnSfx,
        string finishSfx, bool customSound,
        bool startShake, bool nodeShake, bool returnShake,
        bool tickShake, bool permaShake, string syncTag,
        bool hideRope, bool hideCog, bool dashable, bool onlyDash, bool dashableOnce,
        bool dashableRefill, string returnDelays, string returnSpeeds,
        float irrespondTime, string returnEase, bool nodeSound,
        bool unit, bool sideFlag
        )
        : base(data, position, width, height)
    {
        Depth = Depths.FGTerrain + 1;

        this.nodes = nodes;

        this.theme = theme;
        this.permanent = permanent;
        this.waits = waits;
        this.ticking = ticking;

        Add(seq = new Coroutine(Sequence()));
        Add(new LightOcclude());

        SurfaceSoundIndex = SurfaceIndex.Girder;

        string path, id, key, corners;
        if (!string.IsNullOrEmpty(customSkin))
        {
            path = customSkin + "/light";
            id = customSkin + "/block";
            key = customSkin + "/innercog";
            corners = customSkin + "/innerCorners";
            cog = GFX.Game[customSkin + "/cog"];
            themePath = "normal";
            backgroundColor = Color.Black;
            if (this.theme == Themes.Moon)
                themePath = "moon";
        }
        else
        {
            switch (this.theme)
            {
                default:
                case Themes.Normal:
                    path = "objects/zipmover/light";
                    id = "objects/zipmover/block";
                    key = "objects/zipmover/innercog";
                    corners = "objects/ChroniaHelper/zipmover/innerCorners";
                    cog = GFX.Game["objects/zipmover/cog"];
                    themePath = "normal";
                    drawBlackBorder = true;
                    backgroundColor = Color.Black;
                    break;

                case Themes.Moon:
                    path = "objects/zipmover/moon/light";
                    id = "objects/zipmover/moon/block";
                    key = "objects/zipmover/moon/innercog";
                    corners = "objects/ChroniaHelper/zipmover/moon/innerCorners";
                    cog = GFX.Game["objects/zipmover/moon/cog"];
                    themePath = "moon";
                    drawBlackBorder = false;
                    backgroundColor = Color.Black;
                    break;

                case Themes.Cliffside:
                    path = "objects/ChroniaHelper/omniZipMover/cliffside/light";
                    id = "objects/ChroniaHelper/omniZipMover/cliffside/block";
                    key = "objects/ChroniaHelper/omniZipMover/cliffside/innercog";
                    corners = "objects/ChroniaHelper/omniZipMover/cliffside/innerCorners";
                    cog = GFX.Game["objects/ChroniaHelper/omniZipMover/cliffside/cog"];
                    themePath = "normal";
                    drawBlackBorder = true;
                    backgroundColor = Calc.HexToColor("171018");
                    break;
            }
        }

        if (!string.IsNullOrEmpty(bgColor)) { this.backgroundColor = Calc.HexToColor(bgColor); }
        if (!string.IsNullOrEmpty(ropeColor)) { this.ropeColor = Calc.HexToColor(ropeColor); }
        if (!string.IsNullOrEmpty(ropeLightColor)) { this.ropeLightColor = Calc.HexToColor(ropeLightColor); }

        innerCogs = GFX.Game.GetAtlasSubtextures(key);
        streetlight = new Sprite(GFX.Game, path)
        {
            Active = false
        };
        streetlight.Add("frames", "", 1f);
        streetlight.Play("frames");
        streetlight.SetAnimationFrame(1);
        streetlight.Position = new Vector2((Width / 2f) - (streetlight.Width / 2f), 0f);
        Add(bloom = new BloomPoint(1f, 6f)
        {
            Position = new Vector2(Width / 2f, 4f)
        });

        // Check inner corners
        if (!GFX.Game.HasAtlasSubtextures(corners))
        {
            corners = "objects/ChroniaHelper/omniZipMover/empty_preset";
        }

        if (legacyCustomTexture != "")
        {
            Tuple<MTexture[,], MTexture[,]> customTiles = SetupCustomTileset(legacyCustomTexture);
            edges = customTiles.Item1;
            innerCorners = customTiles.Item2;
        }
        else
        {
            for (int i = 0; i < 3; i++)
                for (int j = 0; j < 3; j++)
                    edges[i, j] = GFX.Game[id].GetSubtexture(i * 8, j * 8, 8, 8);

            for (int i = 0; i < 2; i++)
                for (int j = 0; j < 2; j++)
                    innerCorners[i, j] = GFX.Game[corners].GetSubtexture(i * 8, j * 8, 8, 8);
        }

        Add(sfx = new SoundSource()
        {
            Position = new Vector2(Width, Height) / 2f
        });

        // New params

        this.delays = delays.Split(',',StringSplitOptions.TrimEntries);
        this.easing = ease.Split(',',StringSplitOptions.TrimEntries);
        this.speeds = speeds.Split(',', StringSplitOptions.TrimEntries);
        if (ticks <= 0) { ticks = 5; }
        if (tickDelay <= Engine.DeltaTime) { tickDelay = Engine.DeltaTime; }
        this.ticks = ticks;
        this.tickDelay = tickDelay;

        if (this.synced = sync)
        {
            if(syncTag.Trim() == "")
            {
                this.zmtag = ropeColor;
            }
            else
            {
                this.zmtag = syncTag;
            }
        }
        
        this.startDelay = Math.Abs(startdelay);

        this.startSound = startSfx;
        this.impactSound = impactSfx;
        this.tickSound = tickSfx;
        this.returnSound = returnSfx;
        this.finishSound = finishSfx;

        this.customSound = customSound;

        this.startShake = startShake;
        this.tickShake = tickShake;
        this.permaShake = permaShake;
        this.returnShake = returnShake;
        this.nodeShake = nodeShake;
        this.hideRope = hideRope;
        this.hideCog = hideCog;

        // OnDash
        onDash = data.Enum<OnDash>("onDash", OnDash.Normal);
        this.dashable = dashable;
        if (!string.IsNullOrEmpty(data.Attr("dashable"))){
            if (this.dashable) { onDash = OnDash.Rebound; } // old data
        }
        this.onlyDash = onlyDash;
        this.dashableOnce = dashableOnce;
        this.dashableRefill = dashableRefill;
        
        OnDashCollide = OnDashed;
        

        // Return params
        this.returnDelays = returnDelays.Split(',', StringSplitOptions.TrimEntries);
        this.returnSpeeds = returnSpeeds.Split(',', StringSplitOptions.TrimEntries);
        irrespondingTime = irrespondTime;
        returnEasing = returnEase.Split(',',StringSplitOptions.TrimEntries);

        // Node Sound
        this.nodeSound = nodeSound;

        // Time Units
        timeUnits = unit;

        sideflag = sideFlag;

        // flag nodes?
        flagConditions = data.Attr("nodeFlags");
        conditionEmpty = string.IsNullOrEmpty(flagConditions);
        flags = FlagUtils.ParseList(flagConditions);

        // touch sensitive
        sensitive = data.Enum<TouchSensitive>("touchSensitive", TouchSensitive.none);

        // render
        recPos = position;
        recWidth = width;
        recHeight = height;
        
    }


    public override void Awake(Scene scene)
    {
        base.Awake(scene);
        AutoTile(edges, innerCorners);

        Add(streetlight);
    }

    public override void Added(Scene scene)
    {
        base.Added(scene);
        level = SceneAs<Level>();
        scene.Add(pathRenderer = new(this, nodes, ropeColor, ropeLightColor));
    }

    public override void Removed(Scene scene)
    {
        scene.Remove(pathRenderer);
        pathRenderer = null;
        base.Removed(scene);
    }

    public override void Update()
    {
        base.Update();

        bloom.Visible = streetlight.CurrentAnimationFrame != 0;
        bloom.Y = (theme == Themes.Normal) ? streetlight.CurrentAnimationFrame * 3 : (theme == Themes.Cliffside) ? 5 : 9;

        // OnDash
        dcr = DashCollisionResults.NormalCollision;
        scale = Calc.Approach(scale, Vector2.One, 3f * Engine.DeltaTime);

        streetlight.Scale = scale;
        Vector2 zeroCenter = new Vector2(Width, Height) / 2f;
        streetlight.Position = zeroCenter + (new Vector2(zeroCenter.X - streetlight.Width / 2f, 0) - zeroCenter) * scale;
    }

    public override void Render()
    {
        Vector2 originalPosition = Position;
        Position += Shake;

        foreach (Hitbox extension in Colliders)
        {
            if (theme == Themes.Moon)
            {
                Draw.Rect(extension.Left + 2f + X, extension.Top + Y, extension.Width - 4f, extension.Height, backgroundColor);
                Draw.Rect(extension.Left + X, extension.Top + 2f + Y, extension.Width, extension.Height - 4f, backgroundColor);
                foreach (Image t in InnerCornerTiles)
                    Draw.Rect(t.X + X, t.Y + Y, 8, 8, backgroundColor);
            }
            else
                Draw.Rect(extension.Left + X, extension.Top + Y, extension.Width, extension.Height, backgroundColor);
        }

        int n = 1;
        float rot = 0f;
        int count = innerCogs.Count;

        float w = GroupBoundsMax.X - GroupBoundsMin.X;
        float h = GroupBoundsMax.Y - GroupBoundsMin.Y;
        Vector2 offset = new Vector2(-8, -8) + GroupOffset;

        for (int i = 4; i <= h + 4; i += 8)
        {
            int oldN = n;
            for (int j = 4; j <= w + 4; j += 8)
            {
                int index = (int) (NumberUtils.Mod((rot + (n * percent * (float) Math.PI * 4f)) / ((float) Math.PI / 2f), 1f) * count);
                MTexture mTexture = innerCogs[index];
                Rectangle rectangle = new(0, 0, mTexture.Width, mTexture.Height);
                Vector2 zero = Vector2.Zero;

                int x = (j - 4) / 8;
                int y = (i - 4) / 8;
                if (GroupTiles[x, y])
                {
                    // Rescaling the SubTexture Rectangle if the current cog can be rendered outside the Zip Mover

                    if (!GroupTiles[x - 1, y]) // Left
                    {
                        zero.X = 2f;
                        rectangle.X = 2;
                        rectangle.Width -= 2;
                    }
                    if (!GroupTiles[x + 1, y]) // Right
                    {
                        zero.X = -2f;
                        rectangle.Width -= 2;
                    }
                    if (!GroupTiles[x, y - 1]) // Up
                    {
                        zero.Y = 2f;
                        rectangle.Y = 2;
                        rectangle.Height -= 2;
                    }
                    if (!GroupTiles[x, y + 1]) // Down
                    {
                        zero.Y = -2f;
                        rectangle.Height -= 2;
                    }

                    mTexture = mTexture.GetSubtexture(rectangle.X, rectangle.Y, rectangle.Width, rectangle.Height, temp);
                    mTexture.DrawCentered(Position + new Vector2(j, i) + zero + offset, Color.White * ((n < 0) ? 0.5f : 1f));
                }

                n *= -1;
                rot += MathHelper.Pi / 3f;
            }

            // Ensures the checkboard pattern for innercogs
            if (oldN == n)
                n *= -1;
        }

        base.Render();
        Position = originalPosition;
    }

}

File: ModFolder/Loenn/entities/OmniZipMover.lua
local omniSetups = require("mods").requireFromPlugin("consts.omniSetups")
local drawableRectangle = require("structs.drawable_rectangle")
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")
local connectedEntities = require("helpers.connected_entities")
local ChroniaHelper = require("mods").requireFromPlugin("helpers.chroniaHelper")
local fo = require("mods").requireFromPlugin("helpers.field_options")
local drawableText = require("structs.drawable_text")

local OmniZipMover = {}

local themes = {
    "Normal",
    "Moon",
    "Cliffside"
}

OmniZipMover.associatedMods = { "CommunalHelper", "ChroniaHelper" }

OmniZipMover.name = "ChroniaHelper/OmniZipMover"
OmniZipMover.depth = 4999
OmniZipMover.minimumSize = {16, 16}
OmniZipMover.nodeLimits = {1, -1}
OmniZipMover.nodeVisibility = "never"
OmniZipMover.fieldInformation = {
    theme = {
        options = themes,
        editable = false
    },
    backgroundColor = {
        fieldType = "color",
        allowEmpty = true,
    },
    ropeColor = {
        fieldType = "color",
        allowEmpty = true,
    },
    ropeLightColor = {
        fieldType = "color",
        allowEmpty = true,
    },
    easing = {
        fieldType = "list",
        elementOptions = {
            options = ChroniaHelper.easers,
            editable = false,
        },
        allowEmpty = false,
    },
    returnEasing = {
        fieldType = "list",
        elementOptions = {
            options = ChroniaHelper.easers,
            editable = false,
        },
        allowEmpty = false,
    },
    ticks = {
        fieldType = "integer",
    },
    startSound = {
        options = {"event:/CommunalHelperEvents/game/zipMover/normal/start"},
        editable = true,
    },
    impactSound = {
        options = {"event:/CommunalHelperEvents/game/zipMover/normal/impact"},
        editable = true,
    },
    tickSound = {
        options = {"event:/CommunalHelperEvents/game/zipMover/normal/tick"},
        editable = true,
    },
    returnSound = {
        options = {"event:/CommunalHelperEvents/game/zipMover/normal/return"},
        editable = true,
    },
    finishSound = {
        options = {"event:/CommunalHelperEvents/game/zipMover/normal/finish"},
        editable = true,
    },
    customSkin = require("mods").requireFromPlugin("helpers.vivUtilsMig").getDirectoryPathFromFile(true),
    delays = {
        fieldType = "list",
    },
    nodeSpeeds = {
        fieldType = "list",
    },
    returnSpeeds = {
        fieldType = "list",
    },
    returnDelays = {
        fieldType = "list",
    },
    touchSensitive = {
        options = {
            "none", "bottom", "sideways", "always"
        },
        editable = false,
    },
    onDash = {
        options = {
            "Rebound", "Bounce", "Normal",
        },
        editable = false,
    },
    indicator = {
        fieldType = "color",
        allowEmpty = true,
    },
}

OmniZipMover.fieldOrder = omniSetups.fieldOrder

OmniZipMover.placements = {}
for i, theme in ipairs(themes) do
    OmniZipMover.placements[i] = {
        name = string.lower(theme),
        placementType = "rectangle",
        data = {
            width = 16,
            height = 16,
            theme = theme,
            delays = "0.2,0.2",
            permanent = false,
            waiting = false,
            ticking = false,
            synced = false,
            customSkin = "",
            backgroundColor = "",
            ropeColor = "",
            --default 663931
            ropeLightColor = "",
            --default 9b6157
            easing = "SineIn",
            nodeSpeeds = "1,1",
            ticks = 5,
            tickDelay = 1,
            startDelay = 0.1,
            customSound = false,
            startSound = "event:/CommunalHelperEvents/game/zipMover/normal/start",
            impactSound = "event:/CommunalHelperEvents/game/zipMover/normal/impact",
            tickSound = "event:/CommunalHelperEvents/game/zipMover/normal/tick",
            returnSound = "event:/CommunalHelperEvents/game/zipMover/normal/return",
            finishSound = "event:/CommunalHelperEvents/game/zipMover/normal/finish",
            nodeSound = true;
            startShaking = true,
            nodeShaking = true,
            returnShaking = true,
            tickingShaking = true,
            permanentArrivalShaking = true,
            tweakShakes = false,
            syncTag = "",
            hideRope = false,
            hideCog = false,
            -- dashable params
            --dashable = false,
            onDash = "Normal",
            onlyDashActivate = false,
            dashableOnce = true,
            dashableRefill = false,
            -- return params
            tweakReturnParams = false,
            returnSpeeds = "1,1",
            returnDelays = "0.2,0.2",
            returnedIrrespondingTime = 0.5,
            returnEasing = "SineIn",
            -- time
            timeUnits = false,
            sideFlag = false,
            nodeFlags = "",
            touchSensitive = "none",
            displayParameters = false,
            indicator = "",
            indicatorExpansion = 0,
        }
    }
end

OmniZipMover.ignoredFields = function(entity)
    return ignoredAttrs(entity)
end

function ignoredAttrs(entity)
    local attrs = {"_name","_id","x","y"}
    local soundAttrs = {
        "startSound",
        "impactSound","tickSound",
        "returnSound","finishSound","nodeSound",
    }

    local shakeAttrs = {
        "startShaking", "nodeShaking",
        "returnShaking", "tickingShaking",
        "permanentArrivalShaking",
    }

    local returnAttrs = {
        "returnDelays", "returnSpeeds", "returnEasing",
        "returnedIrrespondingTime",
    }

    local dashableAttrs = {
        "dashableOnce", "dashableRefill","onlyDashActivate",
    }

    if not entity.customSound then
        for _, item in ipairs(soundAttrs) do
            table.insert(attrs, item)
        end
    end

    if not entity.tweakShakes then
        for _, item in ipairs(shakeAttrs) do
            table.insert(attrs, item)
        end
    end

    if not entity.tweakReturnParams then
        for _, item in ipairs(returnAttrs) do
            table.insert(attrs, item)
        end
    end

    --[[
    if not entity.dashable then
        for _, item in ipairs(dashableAttrs) do
            table.insert(attrs, item)
        end
    end
    ]]
    

    return attrs
end

local zipMoverRoleColor = {102 / 255, 57 / 255, 49 / 255}

local function getSearchPredicate()
    return function(target)
        return target._name == "ChroniaHelper/SolidExtension"
    end
end

local function getTileSprite(entity, x, y, block, inner, rectangles)
    local hasAdjacent = connectedEntities.hasAdjacent

    local drawX, drawY = (x - 1) * 8, (y - 1) * 8

    local closedLeft = hasAdjacent(entity, drawX - 8, drawY, rectangles)
    local closedRight = hasAdjacent(entity, drawX + 8, drawY, rectangles)
    local closedUp = hasAdjacent(entity, drawX, drawY - 8, rectangles)
    local closedDown = hasAdjacent(entity, drawX, drawY + 8, rectangles)
    local completelyClosed = closedLeft and closedRight and closedUp and closedDown

    local quadX, quadY = false, false
    local frame = block

    if completelyClosed then
        frame = inner
        if not hasAdjacent(entity, drawX + 8, drawY - 8, rectangles) then
            quadX, quadY = 8, 0
        elseif not hasAdjacent(entity, drawX - 8, drawY - 8, rectangles) then
            quadX, quadY = 0, 0
        elseif not hasAdjacent(entity, drawX + 8, drawY + 8, rectangles) then
            quadX, quadY = 8, 8
        elseif not hasAdjacent(entity, drawX - 8, drawY + 8, rectangles) then
            quadX, quadY = 0, 8
        else
            quadX, quadY = 8, 8
            frame = block
        end
    else
        if closedLeft and closedRight and not closedUp and closedDown then
            quadX, quadY = 8, 0
        elseif closedLeft and closedRight and closedUp and not closedDown then
            quadX, quadY = 8, 16
        elseif closedLeft and not closedRight and closedUp and closedDown then
            quadX, quadY = 16, 8
        elseif not closedLeft and closedRight and closedUp and closedDown then
            quadX, quadY = 0, 8
        elseif closedLeft and not closedRight and not closedUp and closedDown then
            quadX, quadY = 16, 0
        elseif not closedLeft and closedRight and not closedUp and closedDown then
            quadX, quadY = 0, 0
        elseif not closedLeft and closedRight and closedUp and not closedDown then
            quadX, quadY = 0, 16
        elseif closedLeft and not closedRight and closedUp and not closedDown then
            quadX, quadY = 16, 16
        end
    end

    if quadX and quadY then
        local sprite = drawableSprite.fromTexture(frame, entity)

        sprite:addPosition(drawX, drawY)
        sprite:useRelativeQuad(quadX, quadY, 8, 8)

        return sprite
    end
end

local function getOmniZipMoverThemeData(entity)
    local theme = string.lower(entity.theme or "normal")
    local cliffside = theme == "cliffside"
    local folder = cliffside and "ChroniaHelper/omniZipMover" or "zipmover"
    local themePath = (theme == "normal") and "" or (theme .. "/")

    local customSkin = entity.customSkin or ""
    if customSkin ~= "" then
        return {
            block = customSkin .. "/block",
            light = customSkin .. "/light01",
            cog = customSkin .. "/cog",
            inner = customSkin .. "/innerCorners"
        }
    end

    return {
        block = "objects/" .. folder .. "/" .. themePath .. "block",
        light = "objects/" .. folder .. "/" .. themePath .. "light01",
        cog = "objects/" .. folder .. "/" .. themePath .. "cog",
        inner = "objects/" .. ((cliffside and "" or "ChroniaHelper/") .. folder) .. "/" .. themePath .. "innerCorners"
    }
end

function OmniZipMover.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 16, entity.height or 16
    local tileWidth, tileHeight = math.ceil(width / 8), math.ceil(height / 8)

    local rectangle = drawableRectangle.fromRectangle("fill", x + 2, y + 2, width - 4, height - 4, require('mods').requireFromPlugin('helpers.vivUtilsMig').getColorTable(entity.backgroundColor, true, {0,0,0,1}))
    local sprites = {rectangle:getDrawableSprite()}

    local themeData = getOmniZipMoverThemeData(entity)

    local nodes = entity.nodes or {{x = 0, y = 0}}
    local nodeSprites = ChroniaHelper.getZipMoverNodeSprites(x, y, width, height, nodes, themeData.cog, {1, 1, 1}, require('mods').requireFromPlugin('helpers.vivUtilsMig').getColorTable(entity.ropeColor, true, {1,1,1,1}))
    for _, sprite in ipairs(nodeSprites) do
        table.insert(sprites, sprite)
    end

    local relevantBlocks = utils.filter(getSearchPredicate(), room.entities)
    connectedEntities.appendIfMissing(relevantBlocks, entity)
    local rectangles = connectedEntities.getEntityRectangles(relevantBlocks)

    for i = 1, tileWidth do
        for j = 1, tileHeight do
            local sprite = getTileSprite(entity, i, j, themeData.block, themeData.inner, rectangles)

            if sprite then
                table.insert(sprites, sprite)
            end
        end
    end

    for nodeIndex, node in ipairs(nodes) do
        for i = 1, tileWidth do
            for j = 1, tileHeight do
                local sprite = getTileSprite(entity, i, j, themeData.block, themeData.inner, rectangles)
                
                if sprite then
                    sprite:addPosition(node.x - entity.x, node.y - entity.y)
                    sprite:setColor({1,1,1,0.3})
                    table.insert(sprites, sprite)
                end
            end
        end

        local text = drawableText.fromText(getStringAtIndex(entity.nodeSpeeds, nodeIndex), node.x, node.y)
        local texta = drawableText.fromText(getStringAtIndex(entity.delays, nodeIndex), node.x, node.y, entity.width * 2, entity.height * 2)
        if entity.displayParameters then
            table.insert(sprites, text)
            table.insert(sprites, texta)
        end
    end

    local lightsSprite = drawableSprite.fromTexture(themeData.light, entity)
    lightsSprite:addPosition(math.floor(width / 2), 0)
    lightsSprite:setJustification(0.5, 0.0)
    table.insert(sprites, lightsSprite)

    return sprites
end

function getStringAtIndex(str, index)
    local strings = {}
    for substring in str:gmatch("([^,]+)") do
        strings[#strings + 1] = substring
    end
    
    if #strings == 0 then
        return ""
    elseif index <= #strings then
        return strings[index]
    else
        return strings[#strings]
    end
end

function OmniZipMover.selection(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 8, entity.height or 8
    local halfWidth, halfHeight = math.floor(entity.width / 2), math.floor(entity.height / 2)

    local mainRectangle = utils.rectangle(x, y, width, height)

    local nodes = entity.nodes or {{x = 0, y = 0}}
    local nodeRectangles = {}
    for _, node in ipairs(nodes) do
        local centerNodeX, centerNodeY = node.x + halfWidth, node.y + halfHeight

        table.insert(nodeRectangles, utils.rectangle(centerNodeX - 5, centerNodeY - 5, 10, 10))
    end

    return mainRectangle, nodeRectangles
end

return OmniZipMover
